src/load_cell_endstop.c

```diff
@@ -0,0 +1,297 @@
+void
+command_config_load_cell_endstop(uint32_t *args)
+{
+    struct load_cell_endstop *lce = oid_alloc(args[0]
+                            , command_config_load_cell_endstop, sizeof(*lce));
+    lce->flags = 0;
+    lce->trigger_ticks = 0;
+    lce->watchdog_max = 0;
+    lce->watchdog_count = 0;
+    lce->sf = sos_filter_oid_lookup(args[1]);
+    set_endstop_range(lce, 0, 0, 0, 0, 0);
+}
+DECL_COMMAND(command_config_load_cell_endstop, "config_load_cell_endstop"
+                                               " oid=%c sos_filter_oid=%c");
+
+// Lookup a load_cell_endstop
+struct load_cell_endstop *
+load_cell_endstop_oid_lookup(uint8_t oid)
+{
+    return oid_lookup(oid, command_config_load_cell_endstop);
+}
+
+// Set the triggering range and tare value
+void
+command_set_range_load_cell_endstop(uint32_t *args)
+{
+    struct load_cell_endstop *lce = load_cell_endstop_oid_lookup(args[0]);
+    set_endstop_range(lce, args[1], args[2], args[3], args[4]
+                , (fixedQ16_t)args[5]);
+}
+DECL_COMMAND(command_set_range_load_cell_endstop, "set_range_load_cell_endstop"
+    " oid=%c safety_counts_min=%i safety_counts_max=%i tare_counts=%i"
+    " trigger_grams=%u grams_per_count=%i");
+
+// Home an axis
+void
+command_load_cell_endstop_home(uint32_t *args)
+{
+    struct load_cell_endstop *lce = load_cell_endstop_oid_lookup(args[0]);
+    sched_del_timer(&lce->time);
+    // clear the homing trigger flag
+    clear_flag(FLAG_IS_HOMING_TRIGGER, lce);
+    clear_flag(FLAG_IS_HOMING, lce);
+    lce->trigger_ticks = 0;
+    lce->ts = NULL;
+    // 0 samples indicates homing is finished
+    if (args[3] == 0) {
+        // Disable end stop checking
+        return;
+    }
+    lce->ts = trsync_oid_lookup(args[1]);
+    lce->trigger_reason = args[2];
+    lce->error_reason = args[3];
+    lce->time.waketime = args[4];
+    lce->rest_ticks = args[5];
+    lce->watchdog_max = args[6];
+    lce->watchdog_count = 0;
+    reset_filter_state(lce->sf);
+    lce->time.func = watchdog_event;
+    sched_add_timer(&lce->time);
+    set_flag(FLAG_IS_HOMING, lce);
+}
+DECL_COMMAND(command_load_cell_endstop_home,
+             "load_cell_endstop_home oid=%c trsync_oid=%c trigger_reason=%c"
+             " error_reason=%c clock=%u rest_ticks=%u"
+             " timeout=%u");
+
+void
+command_load_cell_endstop_query_state(uint32_t *args)
+{
+    uint8_t oid = args[0];
+    struct load_cell_endstop *lce = load_cell_endstop_oid_lookup(args[0]);
+    sendf("load_cell_endstop_state oid=%c homing=%c homing_triggered=%c"
+        " is_triggered=%c trigger_ticks=%u sample=%i sample_ticks=%u"
+            , oid, is_flag_set(FLAG_IS_HOMING, lce)
+            , is_flag_set(FLAG_IS_HOMING_TRIGGER, lce)
+            , is_flag_set(FLAG_IS_TRIGGERED, lce), lce->trigger_ticks
+            , lce->last_sample, lce->last_sample_ticks);
+}
+DECL_COMMAND(command_load_cell_endstop_query_state
+                , "load_cell_endstop_query_state oid=%c");
```

Comment:
I don't think it is a good idea for the mcu loadcell code to try to emulate QUERY_ENDSTOP.  I'd be concerned that it would be an ongoing maintenance burden - as testing and reflashing the mcu is a pain.

I can see the value in providing a diagnostic tool, but I think that tool should be loadcell specific (not QUERY_ENDSTOP nor QUERY_PROBE) and the command should be host based.  The host can start the sensor reports, analyze data, and provide detailed information to the user.  That kind of host based tool is also much easier to maintain.

Reply:
removed, replaced with `LOAD_CELL_TEST_TAP`